File: src/providers.ts
import { PROVIDERS, ProviderName } from './config';
import { mockChat } from './mockResponses';

/** Decide whether to run without network. Default: mock when no keys are set. */
export function isMock(): boolean {
  if (process.env.LUMEN_MOCK === '1') return true;
  if (process.env.LUMEN_MOCK === '0') return false;
  return !Object.values(PROVIDERS).some((p) => p.apiKey);
}

// Lazily created OpenAI-compatible clients, one per provider.
const clients: Partial<Record<ProviderName, any>> = {};

async function clientFor(provider: ProviderName): Promise<any> {
  if (!clients[provider]) {
    const cfg = PROVIDERS[provider];
    if (!cfg.apiKey) {
      throw new Error(`Missing ${cfg.envKey}. Set it in .env, or run in mock mode (leave keys blank).`);
    }
    const { default: OpenAI } = await import('openai');
    clients[provider] = new OpenAI({ baseURL: cfg.baseURL, apiKey: cfg.apiKey });
  }
  return clients[provider];
}

export interface ChatOptions {
  provider: ProviderName;
  model: string;
  system: string;
  user: string;
  temperature?: number;
  json?: boolean;
  /** Stable key the mock backend uses to return canned content for this step. */
  mockKey: string;
}

/**
 * One call to a model. In mock mode this returns deterministic canned content
 * (keyed by mockKey) so the whole pipeline runs with no keys and no network.
 * Both providers are OpenAI-compatible, so the real path is identical.
 */
export async function chat(opts: ChatOptions): Promise<string> {
  if (isMock()) {
    // Optional pacing so the live web room is watchable (real mode is paced by
    // network latency). Set LUMEN_MOCK_DELAY_MS=0 for instant CLI runs.
    const delay = Number(process.env.LUMEN_MOCK_DELAY_MS ?? 0);
    if (delay > 0) await new Promise((r) => setTimeout(r, delay));
    return mockChat(opts.mockKey);
  }

  const client = await clientFor(opts.provider);
  const res = await client.chat.completions.create({
    model: opts.model,
    temperature: opts.temperature ?? 0.2,
    messages: [
      { role: 'system', content: opts.system },
      { role: 'user', content: opts.user },
    ],
    ...(opts.json ? { response_format: { type: 'json_object' } } : {}),
  });
  return res.choices?.[0]?.message?.content ?? '';
}

File: backend/app/providers.py
"""Model provider client + mock switch.

Active defaults use OpenAI + Anthropic through OpenAI-compatible chat endpoints.
Gemini remains in the provider registry only for explicit reassignment. Mock mode
returns deterministic canned content so the whole pipeline runs with no keys/network.
"""
from __future__ import annotations
import asyncio
import os

from .config import PROVIDERS
from .mock_responses import mock_chat


def is_mock() -> bool:
    flag = os.getenv("LUMEN_MOCK")
    if flag == "1":
        return True
    if flag == "0":
        return False
    return not any(p.api_key for p in PROVIDERS.values())


_clients: dict[str, object] = {}


def _client_for(provider: str):
    if provider not in _clients:
        cfg = PROVIDERS[provider]
        if not cfg.api_key:
            raise RuntimeError(
                f"Missing {cfg.env_key}. Set it in .env, or run in mock mode (leave keys blank)."
            )
        from openai import AsyncOpenAI

        _clients[provider] = AsyncOpenAI(base_url=cfg.base_url, api_key=cfg.api_key)
    return _clients[provider]


async def chat(
    *,
    provider: str,
    model: str,
    system: str,
    user: str,
    mock_key: str,
    temperature: float = 0.2,
    json: bool = True,
) -> str:
    if is_mock():
        # Optional pacing so the live web room is watchable.
        delay = float(os.getenv("LUMEN_MOCK_DELAY_MS", "0")) / 1000.0
        if delay > 0:
            await asyncio.sleep(delay)
        return mock_chat(mock_key)

    client = _client_for(provider)
    kwargs = {
        "model": model,
        "messages": [
            {"role": "system", "content": system},
            {"role": "user", "content": user},
        ],
    }
    if provider != "anthropic":
        kwargs["temperature"] = temperature
    if json and provider != "anthropic":
        kwargs["response_format"] = {"type": "json_object"}
    res = await client.chat.completions.create(**kwargs)  # type: ignore[attr-defined]
    return res.choices[0].message.content or ""
